components/cloud/gcp/src/gcs.rs

```diff
@@ -345,6 +364,86 @@ impl BlobStorage for GcsStorage {
     }
 }
 
+struct GcsWalker<'cli, 'arg> {
+    cli: &'cli GcsStorage,
+    page_token: Option<String>,
+    prefix: &'arg str,
+    finished: bool,
+}
+
+impl<'cli, 'arg> GcsWalker<'cli, 'arg> {
+    async fn one_page(&mut self) -> io::Result<Option<Vec<BlobObject>>> {
+        if self.finished {
+            return Ok(None);
+        }
+
+        let mut opt = ListOptional::default();
+        let bucket =
+            BucketName::try_from(self.cli.config.bucket.bucket.to_string()).or_invalid_input(
+                format_args!("invalid bucket {}", self.cli.config.bucket.bucket),
+            )?;
+        let prefix = self.cli.maybe_prefix_key(self.prefix);
+        opt.prefix = Some(&prefix);
+        opt.max_results = Some(2048);
+        opt.page_token = self.page_token.as_deref();
+        let req = Object::list(&bucket, Some(opt)).or_io_error(format_args!(
+            "failed to list with prefix {} page_token {:?}",
+            self.prefix, self.page_token
+        ))?;
+        let res = self
+            .cli
+            .make_request(req.map(|_e| Body::empty()), tame_gcs::Scopes::ReadOnly)
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
+        let resp = utils::read_from_http_body::<ListResponse>(res).await?;
+        debug!("requesting paging GCP"; "prefix" => self.prefix, "page_token" => self.page_token.as_deref(), 
+            "response_size" => resp.objects.len(), "new_page_token" => resp.page_token.as_deref());
+        // GCP returns an empty page token when returning the last page...
+        // We need to break there or we will enter an infinity loop...
+        if resp.page_token.is_none() || resp.objects.is_empty() {
+            self.finished = true;
+        }
+        self.page_token = resp.page_token;
+        let items = resp
+            .objects
+            .into_iter()
+            .map(|v| BlobObject {
+                key: self
+                    .cli
+                    .maybe_strip_prefix_for_string(v.name.unwrap_or_default()),
+            })
+            .collect::<Vec<_>>();
+        Ok(Some(items))
+    }
+}
+
+impl WalkBlobStorage for GcsStorage {
+    fn walk<'c, 'a: 'c, 'b: 'c>(
+        &'a self,
+        prefix: &'b str,
+    ) -> std::pin::Pin<
+        Box<
+            dyn futures_util::stream::Stream<
+                    Item = std::result::Result<cloud::blob::BlobObject, io::Error>,
+                > + 'c,
+        >,
+    > {
+        let walker = GcsWalker {
+            cli: self,
+            page_token: None,
+            prefix,
+            finished: false,
+        };
+        let s = stream::try_unfold(walker, |mut w| async move {
```

Comment:
the impl here seems also dup with the one in S3, not sure if can dedup

Reply:
As `aws` and `gcs` are two different crates, I will prefer that don't try to couple them too tight. (Also you may find that most of the code of `XxxWalker` is specify to requesting one page of the storage.) Perhaps we can make a combinator like `generate_with_state`, but I'm afraid that cannot reduce many line of codes...